File: starter-code/session-6-java-module/src/test/java/com/megabank/fraud/service/FraudDetectionServiceTest.java
package com.megabank.fraud.service;

import com.megabank.fraud.model.*;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.DisplayName;
import org.springframework.boot.test.context.SpringBootTest;

import java.math.BigDecimal;
import java.time.LocalDateTime;

import static org.junit.jupiter.api.Assertions.*;

@SpringBootTest
@DisplayName("Fraud Detection Service Tests")
class FraudDetectionServiceTest {
    
    private FraudDetectionService fraudDetectionService;
    
    @BeforeEach
    void setUp() {
        fraudDetectionService = new FraudDetectionService();
    }
    
    @Test
    @DisplayName("Should approve low-risk transaction")
    void shouldApproveLowRiskTransaction() {
        // Given
        Transaction transaction = createBasicTransaction("TXN001", "ACC001", new BigDecimal("100.00"));
        
        // When
        FraudAnalysisResult result = fraudDetectionService.analyzeTransaction(transaction);
        
        // Then
        assertNotNull(result);
        assertEquals("TXN001", result.getTransactionId());
        assertEquals(RiskLevel.LOW, result.getRiskLevel());
        assertEquals("APPROVE", result.getDecision());
        assertTrue(result.getRiskScore() < 0.3);
    }
    
    @Test
    @DisplayName("Should flag high-amount transaction as high risk")
    void shouldFlagHighAmountTransactionAsHighRisk() {
        // Given
        Transaction transaction = createBasicTransaction("TXN002", "ACC002", new BigDecimal("15000.00"));
        
        // When
        FraudAnalysisResult result = fraudDetectionService.analyzeTransaction(transaction);
        
        // Then
        assertNotNull(result);
        assertTrue(result.getRiskScore() >= 0.3, "High amount should increase risk score");
        assertTrue(result.getRiskFactors().stream()
            .anyMatch(factor -> factor.contains("High transaction amount")));
    }
    
    @Test
    @DisplayName("Should detect high-risk country transactions")
    void shouldDetectHighRiskCountryTransactions() {
        // Given
        Transaction transaction = createBasicTransaction("TXN003", "ACC003", new BigDecimal("500.00"));
        transaction.setLocationCountry("XX"); // High-risk country
        
        // When
        FraudAnalysisResult result = fraudDetectionService.analyzeTransaction(transaction);
        
        // Then
        assertTrue(result.getRiskFactors().stream()
            .anyMatch(factor -> factor.contains("high-risk country")));
    }
    
    @Test
    @DisplayName("Should detect unusual hours transactions")
    void shouldDetectUnusualHoursTransactions() {
        // Given
        Transaction transaction = createBasicTransaction("TXN004", "ACC004", new BigDecimal("200.00"));
        transaction.setTransactionTimestamp(LocalDateTime.now().withHour(3)); // 3 AM
        
        // When
        FraudAnalysisResult result = fraudDetectionService.analyzeTransaction(transaction);
        
        // Then
        assertTrue(result.getRiskFactors().stream()
            .anyMatch(factor -> factor.contains("unusual hours")));
    }
    
    @Test
    @DisplayName("Should detect high-risk merchant categories")
    void shouldDetectHighRiskMerchantCategories() {
        // Given
        Transaction transaction = createBasicTransaction("TXN005", "ACC005", new BigDecimal("1000.00"));
        transaction.setMerchantCategory("gambling");
        
        // When
        FraudAnalysisResult result = fraudDetectionService.analyzeTransaction(transaction);
        
        // Then
        assertTrue(result.getRiskFactors().stream()
            .anyMatch(factor -> factor.contains("High-risk merchant category")));
    }
    
    @Test
    @DisplayName("Should detect new device risk")
    void shouldDetectNewDeviceRisk() {
        // Given
        Transaction transaction = createBasicTransaction("TXN006", "ACC006", new BigDecimal("300.00"));
        transaction.setDeviceFingerprint("new-device-123");
        
        // When
        FraudAnalysisResult result = fraudDetectionService.analyzeTransaction(transaction);
        
        // Then
        assertTrue(result.getRiskFactors().stream()
            .anyMatch(factor -> factor.contains("new device")));
    }
    
    @Test
    @DisplayName("Should detect suspicious IP addresses")
    void shouldDetectSuspiciousIPAddresses() {
        // Given
        Transaction transaction = createBasicTransaction("TXN007", "ACC007", new BigDecimal("250.00"));
        transaction.setIpAddress("10.0.0.1"); // Suspicious IP pattern
        
        // When
        FraudAnalysisResult result = fraudDetectionService.analyzeTransaction(transaction);
        
        // Then
        assertTrue(result.getRiskFactors().stream()
            .anyMatch(factor -> factor.contains("suspicious IP")));
    }
    
    @Test
    @DisplayName("Should track transaction velocity and flag multiple transactions")
    void shouldTrackTransactionVelocity() {
        // Given
        String accountId = "ACC008";
        
        // Simulate multiple transactions for same account
        for (int i = 1; i <= 5; i++) {
            Transaction transaction = createBasicTransaction("TXN" + (100 + i), accountId, new BigDecimal("100.00"));
            fraudDetectionService.analyzeTransaction(transaction);
        }
        
        // When - 6th transaction
        Transaction sixthTransaction = createBasicTransaction("TXN106", accountId, new BigDecimal("100.00"));
        FraudAnalysisResult result = fraudDetectionService.analyzeTransaction(sixthTransaction);
        
        // Then
        assertNotNull(result);
        // Should have some velocity-based risk due to multiple transactions
        assertTrue(result.getRiskScore() > 0.0);
    }
    
    @Test
    @DisplayName("Should generate appropriate recommendations based on risk level")
    void shouldGenerateAppropriateRecommendations() {
        // Given - High risk transaction
        Transaction highRiskTransaction = createBasicTransaction("TXN008", "ACC008", new BigDecimal("20000.00"));
        highRiskTransaction.setLocationCountry("XX");
        highRiskTransaction.setMerchantCategory("gambling");
        
        // When
        FraudAnalysisResult result = fraudDetectionService.analyzeTransaction(highRiskTransaction);
        
        // Then
        assertEquals(RiskLevel.HIGH, result.getRiskLevel());
        assertFalse(result.getRecommendations().isEmpty());
        assertTrue(result.getRecommendations().stream()
            .anyMatch(rec -> rec.contains("Manual review")));
    }
    
    @Test
    @DisplayName("Should handle multiple risk factors correctly")
    void shouldHandleMultipleRiskFactorsCorrectly() {
        // Given - Transaction with multiple risk factors
        Transaction multiRiskTransaction = createBasicTransaction("TXN009", "ACC009", new BigDecimal("12000.00"));
        multiRiskTransaction.setLocationCountry("XX");
        multiRiskTransaction.setMerchantCategory("cryptocurrency");
        multiRiskTransaction.setTransactionTimestamp(LocalDateTime.now().withHour(2));
        multiRiskTransaction.setDeviceFingerprint("suspicious-device");
        multiRiskTransaction.setIpAddress("10.0.0.99");
        
        // When
        FraudAnalysisResult result = fraudDetectionService.analyzeTransaction(multiRiskTransaction);
        
        // Then
        assertEquals(RiskLevel.HIGH, result.getRiskLevel());
        assertEquals("DECLINE", result.getDecision());
        assertTrue(result.getRiskFactors().size() >= 4, "Should detect multiple risk factors");
        assertTrue(result.getRiskScore() > 0.7, "Multiple risk factors should result in high score");
    }
    
    @Test
    @DisplayName("Risk level categorization should be correct")
    void riskLevelCategorizationShouldBeCorrect() {
        // Test Low Risk
        assertEquals(RiskLevel.LOW, RiskLevel.fromScore(0.1));
        assertEquals(RiskLevel.LOW, RiskLevel.fromScore(0.29));
        
        // Test Medium Risk
        assertEquals(RiskLevel.MEDIUM, RiskLevel.fromScore(0.3));
        assertEquals(RiskLevel.MEDIUM, RiskLevel.fromScore(0.5));
        assertEquals(RiskLevel.MEDIUM, RiskLevel.fromScore(0.69));
        
        // Test High Risk
        assertEquals(RiskLevel.HIGH, RiskLevel.fromScore(0.7));
        assertEquals(RiskLevel.HIGH, RiskLevel.fromScore(0.9));
        assertEquals(RiskLevel.HIGH, RiskLevel.fromScore(1.0));
    }
    
    private Transaction createBasicTransaction(String transactionId, String accountId, BigDecimal amount) {
        Transaction transaction = new Transaction();
        transaction.setTransactionId(transactionId);
        transaction.setAccountId(accountId);
        transaction.setAmount(amount);
        transaction.setCurrency("USD");
        transaction.setMerchantId("MERCHANT_001");
        transaction.setMerchantCategory("retail");
        transaction.setTransactionTimestamp(LocalDateTime.now());
        transaction.setTransactionType(TransactionType.PURCHASE);
        transaction.setLocationCountry("US");
        transaction.setLocationCity("New York");
        transaction.setIsCardPresent(true);
        return transaction;
    }
}

File: starter-code/session-6-go-sample/internal/service/fraud_detection_service_test.go
package service

import (
	"fmt"
	"fraud-detection-service/internal/models"
	"strings"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
)

func TestFraudDetectionService_AnalyzeTransaction(t *testing.T) {
	service := NewFraudDetectionService()

	t.Run("Should approve low-risk transaction", func(t *testing.T) {
		// Given
		transaction := createBasicTransaction("TXN001", "ACC001", 100.00)

		// When
		result, err := service.AnalyzeTransaction(transaction)

		// Then
		assert.NoError(t, err)
		assert.NotNil(t, result)
		assert.Equal(t, "TXN001", result.TransactionID)
		assert.Equal(t, models.LowRisk, result.RiskLevel)
		assert.Equal(t, "APPROVE", result.Decision)
		assert.True(t, result.RiskScore < 0.3)
	})

	t.Run("Should flag high-amount transaction as high risk", func(t *testing.T) {
		// Given
		transaction := createBasicTransaction("TXN002", "ACC002", 15000.00)

		// When
		result, err := service.AnalyzeTransaction(transaction)

		// Then
		assert.NoError(t, err)
		assert.NotNil(t, result)
		assert.True(t, result.RiskScore >= 0.3, "High amount should increase risk score")
		assert.Contains(t, result.RiskFactors, "High transaction amount")
	})

	t.Run("Should detect high-risk country transactions", func(t *testing.T) {
		// Given
		transaction := createBasicTransaction("TXN003", "ACC003", 500.00)
		transaction.LocationCountry = "XX" // High-risk country

		// When
		result, err := service.AnalyzeTransaction(transaction)

		// Then
		assert.NoError(t, err)
		assert.True(t, containsRiskFactor(result.RiskFactors, "high-risk country"))
	})

	t.Run("Should detect unusual hours transactions", func(t *testing.T) {
		// Given
		transaction := createBasicTransaction("TXN004", "ACC004", 200.00)
		// Set to 3 AM
		transaction.TransactionTimestamp = time.Date(2024, 7, 28, 3, 0, 0, 0, time.UTC)

		// When
		result, err := service.AnalyzeTransaction(transaction)

		// Then
		assert.NoError(t, err)
		assert.True(t, containsRiskFactor(result.RiskFactors, "unusual hours"))
	})

	t.Run("Should detect high-risk merchant categories", func(t *testing.T) {
		// Given
		transaction := createBasicTransaction("TXN005", "ACC005", 1000.00)
		transaction.MerchantCategory = "gambling"

		// When
		result, err := service.AnalyzeTransaction(transaction)

		// Then
		assert.NoError(t, err)
		assert.True(t, containsRiskFactor(result.RiskFactors, "High-risk merchant category"))
	})

	t.Run("Should detect new device risk", func(t *testing.T) {
		// Given
		transaction := createBasicTransaction("TXN006", "ACC006", 300.00)
		transaction.DeviceFingerprint = "new-device-123"

		// When
		result, err := service.AnalyzeTransaction(transaction)

		// Then
		assert.NoError(t, err)
		assert.True(t, containsRiskFactor(result.RiskFactors, "new device"))
	})

	t.Run("Should detect suspicious IP addresses", func(t *testing.T) {
		// Given
		transaction := createBasicTransaction("TXN007", "ACC007", 250.00)
		transaction.IPAddress = "10.0.0.1" // Suspicious IP pattern

		// When
		result, err := service.AnalyzeTransaction(transaction)

		// Then
		assert.NoError(t, err)
		assert.True(t, containsRiskFactor(result.RiskFactors, "suspicious IP"))
	})

	t.Run("Should generate appropriate recommendations based on risk level", func(t *testing.T) {
		// Given - High risk transaction
		transaction := createBasicTransaction("TXN008", "ACC008", 20000.00)
		transaction.LocationCountry = "XX"
		transaction.MerchantCategory = "gambling"

		// When
		result, err := service.AnalyzeTransaction(transaction)

		// Then
		assert.NoError(t, err)
		assert.Equal(t, models.HighRisk, result.RiskLevel)
		assert.NotEmpty(t, result.Recommendations)
		assert.True(t, containsRecommendation(result.Recommendations, "Manual review"))
	})

	t.Run("Should handle multiple risk factors correctly", func(t *testing.T) {
		// Given - Transaction with multiple risk factors
		transaction := createBasicTransaction("TXN009", "ACC009", 12000.00)
		transaction.LocationCountry = "XX"
		transaction.MerchantCategory = "cryptocurrency"
		transaction.TransactionTimestamp = time.Date(2024, 7, 28, 2, 0, 0, 0, time.UTC)
		transaction.DeviceFingerprint = "suspicious-device"
		transaction.IPAddress = "10.0.0.99"

		// When
		result, err := service.AnalyzeTransaction(transaction)

		// Then
		assert.NoError(t, err)
		assert.Equal(t, models.HighRisk, result.RiskLevel)
		assert.Equal(t, "DECLINE", result.Decision)
		assert.True(t, len(result.RiskFactors) >= 4, "Should detect multiple risk factors")
		assert.True(t, result.RiskScore > 0.7, "Multiple risk factors should result in high score")
	})
}

func TestRiskLevel_FromScore(t *testing.T) {
	tests := []struct {
		score    float64
		expected models.RiskLevel
	}{
		{0.1, models.LowRisk},
		{0.29, models.LowRisk},
		{0.3, models.MediumRisk},
		{0.5, models.MediumRisk},
		{0.69, models.MediumRisk},
		{0.7, models.HighRisk},
		{0.9, models.HighRisk},
		{1.0, models.HighRisk},
	}

	for _, tt := range tests {
		t.Run(fmt.Sprintf("score_%.2f", tt.score), func(t *testing.T) {
			result := models.FromScore(tt.score)
			assert.Equal(t, tt.expected, result)
		})
	}
}

func createBasicTransaction(transactionID, accountID string, amount float64) models.Transaction {
	return models.Transaction{
		TransactionID:        transactionID,
		AccountID:            accountID,
		Amount:               amount,
		Currency:             "USD",
		MerchantID:           "MERCHANT_001",
		MerchantCategory:     "retail",
		TransactionTimestamp: time.Now(),
		TransactionType:      models.Purchase,
		LocationCountry:      "US",
		LocationCity:         "New York",
		IsCardPresent:        boolPtr(true),
	}
}

func containsRiskFactor(factors []string, keyword string) bool {
	for _, factor := range factors {
		if strings.Contains(strings.ToLower(factor), strings.ToLower(keyword)) {
			return true
		}
	}
	return false
}

func containsRecommendation(recommendations []string, keyword string) bool {
	for _, rec := range recommendations {
		if strings.Contains(strings.ToLower(rec), strings.ToLower(keyword)) {
			return true
		}
	}
	return false
}

func boolPtr(b bool) *bool {
	return &b
}